include/flatbuffers/flatbuffers.h

```diff
@@ -557,15 +562,56 @@ class vector_downward {
 
  private:
   // You shouldn't really be copying instances of this class.
-  vector_downward(const vector_downward &);
-  vector_downward &operator=(const vector_downward &);
+  vector_dir(const vector_dir &);
+  vector_dir &operator=(const vector_dir &);
 
   size_t reserved_;
   uint8_t *buf_;
   uint8_t *cur_;  // Points at location between empty (below) and used (above).
   const simple_allocator &allocator_;
 };
 
+template<> inline void
+vector_dir<UPWARD>::clear() {
+  if (buf_ == nullptr)
+    buf_ = allocator_.allocate(reserved_);
+
+  cur_ = buf_;
+}
+
+template <> inline uoffset_t
+vector_dir<UPWARD>::size() const {
+  assert(cur_ != nullptr && buf_ != nullptr);
+  return static_cast<uoffset_t>(cur_ - buf_);
+}
+
+template <> inline uint8_t *
+vector_dir<UPWARD>::make_space(size_t len) {
+  if (len > static_cast<size_t>(cur_ - buf_)) {
+    auto old_size = size();
+    auto largest_align = AlignOf<largest_scalar_t>();
```

Comment:
A lot of this code is the same as for the downward case, would need to be refactored to have less duplicate code.